### PR TITLE
feat: adabelief optimizer

### PR DESCRIPTION
**Pull request by fantes**
this PR adds support for adabelief, a mod over adam optimizer https://arxiv.org/abs/2010.07468
(it also selects recitified version of adam by default)

### CODE REVIEW

File: src/caffe/solvers/adam_solver.cu [hunk 1]
```diff
@@ -6,12 +6,15 @@ namespace caffe {
...
+                           bool amsgrad, bool rectified, bool adabelief) {
   CUDA_KERNEL_LOOP(i, N) {
     float gi = g[i];
     float mi = m[i] = m[i]*beta1 + gi*(1-beta1);
     float vi_old = v[i];
-    float vi = v[i] = v[i]*beta2 + gi*gi*(1-beta2);
+    float vi;
+    if (adabelief)
+      gi -= mi;
+    vi = v[i] = v[i]*beta2 + gi*gi*(1-beta2);
```
Comment (beniz) [inline src/caffe/solvers/adam_solver.cu:17]: define `float vi` here instead of uninitialized above ?

File: src/caffe/solvers/adam_solver.cu [hunk 2]
```diff
@@ -43,12 +46,15 @@ __global__ void AdamUpdate(int N, int t, Dtype* g, Dtype* m, Dtype* v,
...
+                                        bool amsgrad, bool rectified, bool adabelief) {
     CUDA_KERNEL_LOOP(i, N) {
       float gi = g[i];
       float mi = m[i] = m[i]*beta1 + gi*(1-beta1);
       float vi_old = v[i];
-      float vi = v[i] = v[i]*beta2 + gi*gi*(1-beta2);
+      float vi;
+      if (adabelief)
+        gi -= mi;
+      vi = v[i] = v[i]*beta2 + gi*gi*(1-beta2);
```
Comment (beniz) [inline src/caffe/solvers/adam_solver.cu:57]: same as above, maybe `float vi` here.

File: src/caffe/solvers/adam_solver.cu (hunk 1 above)
Comment (fantes) [reply src/caffe/solvers/adam_solver.cu:17]: ok done

File: src/caffe/solvers/adam_solver.cu (hunk 2 above)
Comment (fantes) [reply src/caffe/solvers/adam_solver.cu:57]: ok done

**Review by beniz — APPROVED** [review]

**Comment by beniz** [comment]
Great PR, thanks!

**Review by sileht — APPROVED** [review]

**Pull request merged**